### ISSUE TITLE
InternalLogger locks its own logging file

### ISSUE THREAD

**Issue by BasPuls2017** [NLog/NLog#1924]
Hi.
Maybe i am doing something wrong (hopefully), but i don't think so.

We are using NLog with multiple targets and on multiple threads and the internal logger to log exceptions which could occure in our custom target implementation to a file. While testing this implementation we encountered the following error:

>  VS2015 Output:
>  TearDown : System.IO.IOException : The process cannot access the file "C:\logs\internalException.log" because it is being used by another process.

Our configuration looks like this:

```c#
public class LoggerFactory : ILoggerFactory<LoggerConfigurationBase>
{
	public Contracts.ILogger Create(LoggerConfigurationBase loggerConfig, InternalLogConfiguration internalConfiguration)
    {        
        TargetWithLayout target = loggerConfig.GetTarget();
            
        ConfigurationItemFactory.Default.LayoutRenderers.RegisterDefinition("LogEntryLayoutRenderer", typeof(LogEntryLayoutRenderer));
        target.Layout = "${LogEntryLayoutRenderer}";
    
        LoggingRule loggingRule = new LoggingRule(loggerConfig.LoggerName, LogLevel.Info, target);
    
        LogManager.Configuration.LoggingRules.Add(loggingRule);
    
        LogManager.ReconfigExistingLoggers();
        if (internalConfiguration != null)
        {
            LogManager.ThrowExceptions = internalConfiguration.ThrowExceptions;
            InternalLogger.LogFile = internalConfiguration.Filename;
            InternalLogger.LogLevel = this.ConvertFromSeverity(internalConfiguration.Severity);
            InternalLogger.LogToTrace = internalConfiguration.LogToTrace;
        }
    
        return new ViaLogger(LogManager.GetLogger(loggerConfig.LoggerName));
    }
}
```

The Test we are looking like this:
```c#
[TestFixture]
[ExcludeFromCodeCoverage]
[Category(TestCategories.IntegrationTest)]
[Category("ExceptionHandlingTest")]
[Parallelizable(ParallelScope.None)]
public class ExceptionHandlingTests
{
    private readonly String internalExceptionLogFile = "C:/logs/internalException.log";

	[OneTimeSetUp]
    public void ClassSetUp()
    {
        if (File.Exists(this.internalExceptionLogFile))
        {
            File.Delete(this.internalExceptionLogFile);
        }
    }

    [OneTimeTearDown]
    public void ClassTearDown()
    {
        if (File.Exists(this.internalExceptionLogFile))
        {
            //System.IO.IOException : The process cannot access the file "Disk\Folder\internalException.log" because it is being used by another process.
            File.Delete(this.internalExceptionLogFile);
        }
    }

    [SetUp]
    public void TestSetup()
    {
        Thread.CurrentThread.CurrentUICulture = new CultureInfo("en-US");
    }
	
	[Test]
	[TestCase(TestName = "Log one entry into text file with invalid path should log exception into internal log")]
	//Test Failed: InternalLogFile contains not expected error message. Expected: True, But was:  False
	public void LogTextFile_PathNotexists_ShouldWriteExceptionInInternalTextFile()
	{
		LoggerFactory textFileFactory = new LoggerFactory();
		ILogger textFilelogger = textFileFactory.Create(
			new TextfileConfiguration(
				"ExceptionHandlingTextFileLoggerWithInternalLogFile",
				"ExceptionHandlingTestTextFileTarget",
				"ABC:/Path_Not_Exists/ExceptionHandlingTest.txt"),
			new InternalLogConfiguration()
			{
				Filename = this.internalExceptionLogFile,
				LogToTrace = true,
				Severity = Severities.Trace
			}
		);
		textFilelogger.Log(Severities.Info, "TextFile TestMessage");
		FileAssert.Exists(this.internalExceptionLogFile);
	
		String fileContent = new StreamReader(this.internalExceptionLogFile).ReadToEnd();
		String expectedErrorMessage = "Error Error has been raised. Exception: System.NotSupportedException: The given path's format is not supported.";
		Assert.That(fileContent.Contains(expectedErrorMessage), Is.True, "InternalLogFile contains not expected error message");
	}
	
	[Test]
	[TestCase(TestName = "Log one entry into elastic search with invalid url should log exception into internal log")]
	//Test passes
	public void LogElasticSearch_UrlNotexists_ShouldWriteExceptionInInternalTextFile()
	{
		Uri[] nodeUris =
		{
			new Uri("https://domain-not.exists")
		};
		LoggerFactory elasticSearchfactory = new LoggerFactory();
		ILogger elasticSearchlogger = elasticSearchfactory.Create(
			new ElasticsearchConfiguration(
				"ExceptionHandlingElasticSearchLoggerWithInternalLogFile",
				"ExceptionHandlingTestElasticTarget",
				"ExceptionHandlingTest",
				nodeUris),
			new InternalLogConfiguration()
			{
				Filename = this.internalExceptionLogFile,
				LogToTrace = true
			}
		);
		elasticSearchlogger.Log(Severities.Info, "ElasticSearch TestMessage");
		FileAssert.Exists(this.internalExceptionLogFile);
	
		String fileContent = new StreamReader(this.internalExceptionLogFile).ReadToEnd();
		String expectedErrorMessage = "PingFailure: Node: https://domain-not.exists/ Exception: PipelineException";
		Assert.That(fileContent.Contains(expectedErrorMessage), Is.True, "InternalLogFile contains not expected error message");
	}
}
```

As you can see we are using nunit to perform our tests. The issue only happen when we run these to test together.

In my understanding the file used by the internal logger should be closed after an logentry is written and the filehandle should be released, which it doesnt't do.

My question now: Am i doing something wrong? If no, is there a solution/workaround for this issue?

**Comment by 304NotModified** [comment]
What the NLog version? (Nuget version)

**Comment by BasPuls2017** [comment]
We are using v4.4.1 with .net 4.5.2

**Comment by 304NotModified** [comment]
> My question now: Am i doing something wrong?

I think it's a bug. There is also no flush/close method for the InternalLogger.



> there a solution/workaround for this issue?

Don't use log-to-file but use the `InternalLogger.LogWriter property`

**Comment by BasPuls2017** [comment]
Ok. I tried to change the usage of the log file
 `InternalLogger.LogFile = internalConfiguration.Filename;`
to utilize the logwriter
`InternalLogger.LogWriter = new StreamWriter(internalConfiguration.Filename);`
but now both test 
`LogTextFile_PathNotExists_ShouldWriteExceptionInInternalTextFile`
and
`LogElasticSearch_UrlNotexists_ShouldWriteExceptionInInternalTextFile`
fail with the message

>System.IO.IOException : The process cannot access the file 'C:\logs\internalException.log' because it is being used by another process.

**Comment by snakefoot** [comment]
Had a similiar issue when debugging the ConcurrentFileTargetTests. I just made sure that each test-process used an unique log-filename for the InternalLogger.

**Comment by snakefoot** [comment]
> but now both test fail with the message (StreamWriter)

Why not use an unique filename for the InternalLogConfiguration.Filename (or to the StreamWriter) ?
- LogTextFile_PathNotexists_ShouldWriteExceptionInInternalTextFile.txt
- LogElasticSearch_UrlNotexists_ShouldWriteExceptionInInternalTextFile.txt

**Comment by 304NotModified** [comment]
Fixed by #1972 / NLog 4.4.2, isn't it?

**Comment by snakefoot** [comment]
@304NotModified Not if the unit-tests runs in parallel in different unit-test-executor-processes.

But using InternalLogger.LogWriter together with a StringWriter will solve the problem.

**Comment by 304NotModified** [comment]
I think using internal logger with a LogWriter is fine in unit tests. Writing to file in parallel is indeed not supported for the internal logger

**Comment by 304NotModified** [comment]
Closing this as it's a wontfix
